python-checks/src/main/java/org/sonar/python/checks/tests/NotDiscoverableTestMethodCheck.java

```diff
@@ -0,0 +1,130 @@
+import org.sonar.plugins.python.api.symbols.ClassSymbol;
+import org.sonar.plugins.python.api.symbols.Symbol;
+import org.sonar.plugins.python.api.tree.ClassDef;
+import org.sonar.plugins.python.api.tree.FunctionDef;
+import org.sonar.plugins.python.api.tree.Statement;
+import org.sonar.plugins.python.api.tree.Token;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.python.tree.TreeUtils;
+
+import static java.util.stream.Stream.concat;
+
+@Rule(key = "S5899")
+public class NotDiscoverableTestMethodCheck extends PythonSubscriptionCheck {
+
+  private static final String MESSAGE = "Rename this method so that it starts with \"test\" or remove this unused helper.";
+  // All methods of unittest https://docs.python.org/3/library/unittest.html#unittest.TestCase
+  private static final List<String> UNITTEST_RUN_METHODS = List.of("setUp", "tearDown", "setUpClass", "tearDownClass", "run", "skiptTest",
+    "subTest", "debug");
+  private static final List<String> UNITTEST_CHECK_METHODS = List.of("assertEqual", "assertNotEqual", "assertTrue", "assertFalse",
+    "assertIs", "assertIsNot", "assertIsNone", "assertIsNotNone", "assertIn", "assertNotIn", "assertIsInstance", "assertNotIsInstance",
+    "assertRaises", "assertRaisesRegex", "assertWarns", "assertWarnsRegex", "assertLogs", "assertNoLogs", "assertAlmostEqual", "assertGreater",
+    "assertGreaterEqual", "assertLess", "assertLessEqual", "assertRegex", "assertNotRegex", "assertCountEqual", "addTypeEqualityFunc",
+    "fail", "failureException", "longMessage", "maxDiff");
+  private static final List<String> UNITTEST_GATHER_INFO = List.of("countTestCases", "defaultTestResult", "id", "shortDescription", "addCleanup",
+    "doCleanups", "addClassCleanup", "doClassCleanups");
+  private static final List<String> UNITTEST_METHODS = concat(concat(UNITTEST_CHECK_METHODS.stream(), UNITTEST_RUN_METHODS.stream()), UNITTEST_GATHER_INFO.stream())
+    .collect(Collectors.toList());
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.CLASSDEF, ctx -> {
+      ClassDef classDefinition = (ClassDef) ctx.syntaxNode();
+
+      if (inheritsOnlyFromUnitTest(classDefinition)) {
+        HashMap<String, List<Tree>> functionOccurrences = new HashMap<>();
+        List<FunctionDef> allDefinitions = new ArrayList<>();
+        // Assuming there is no definition of function in another one
+        for (Statement statement : classDefinition.body().statements()) {
+          if (statement.is(Tree.Kind.FUNCDEF)) {
+            FunctionDef functionDef = ((FunctionDef) statement);
+            String nameFunction = functionDef.name().name();
+            // If it doesn't override existing methods and doesn't start with test, we add it to the map
+            if (!overrideExistingMethod(nameFunction) && !nameFunction.startsWith("test")) {
+              addValueToMap(functionOccurrences, functionDef.name(), nameFunction);
```

Comment:
I'd suggest to do something like this instead:
```
Symbol symbol = functionDef.name().symbol();
// If it doesn't override existing methods and doesn't start with test, we add it to the map
if (!overrideExistingMethod(nameFunction) && !nameFunction.startsWith("test") && symbol != null && symbol.is(Symbol.Kind.FUNCTION)) {
    suspiciousFunctionsAndDefinitions.put(((FunctionSymbol) symbol), functionDef);
}
```
(Maybe the checks on the symbol can be done separately)